examples/hpc/dla-driver/src/lib.rs

```diff
@@ -18,291 +100,493 @@ macro_rules! set_bits {
-        reg,
-        1
-    );
-    reg = set_bits!(
-        DLA_HANDSHAKE_MAC_ENABLE_OFFSET,
-        DLA_HANDSHAKE_MAC_ENABLE_BITMASK,
-        reg,
-        1
-    );
-    reg = set_bits!(
-        DLA_HANDSHAKE_BYPASS_ENABLE_OFFSET,
-        DLA_HANDSHAKE_BYPASS_ENABLE_BITMASK,
-        reg,
-        1
-    );
-    dla_write_reg(DLA_HANDSHAKE, reg);
-
-    dla_set_kernel_size(1, 3, 3);
-    dla_set_input_size(1, 5, 5);
-
-    dla_set_input_data_bank(0);
-    dla_set_kernel_data_bank(8);
-
-    dla_enable_relu(true);
-    dla_enable_bias(true);
-
-    let mut A = [
-        1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 3, 3, 3, 3, 3, 4, 4, 4, 4, 4, 5, 5, 5, 5, 5,
-    ];
-    let mut B = [0, 1, 0, 1, 1, 1, 0, 1, 0];
-    dla_write_data_bank(MEMORY_BANK_0_OFFSET, &mut A);
-    dla_write_data_bank(MEMORY_BANK_8_OFFSET, &mut B);
-
-    dla_kernel_data_ready(true);
-    dla_input_data_ready(true);
+    pub fn set_simd_select(&mut self, mode: SimdBitMode) {
+        self.simd_mode = mode.clone();
+        let mut reg = self.read_reg(DLA_MAC_CTRL);
+        reg = set_bits!(
+            DLA_SIMD_SELECT_OFFSET,
+            DLA_SIMD_SELECT_BITMASK,
+            reg,
+            mode as usize
+        );
+        self.write_reg(DLA_MAC_CTRL, reg)
+    }
+
+    pub fn get_simd_format(&self) -> SimdBitMode {
+        let mut reg = self.read_reg(DLA_MAC_CTRL);
+        reg = get_bits!(DLA_SIMD_SELECT_BITMASK, reg);
+        match reg {
+            0 => SimdBitMode::EightBits,
+            1 => SimdBitMode::FourBits,
+            2 => SimdBitMode::TwoBits,
+            _ => SimdBitMode::EightBits,
+        }
+    }
+
+    pub fn set_mac_clip(&self, clip_amount: usize) {
+        let mut reg = self.read_reg(DLA_MAC_CTRL);
+        // Cap clipping amount
+        if clip_amount > 21 {
+            reg = set_bits!(DLA_MAC_CLIP_OFFSET, DLA_MAC_CLIP_BITMASK, reg, 0x1F);
+        } else {
+            reg = set_bits!(DLA_MAC_CLIP_OFFSET, DLA_MAC_CLIP_BITMASK, reg, clip_amount);
+        }
+        self.write_reg(DLA_MAC_CTRL, reg)
+    }
+
+    pub fn set_pp_clip(&self, clip_amount: usize) {
+        let mut reg = self.read_reg(DLA_PP_CTRL);
+        // Cap clipping amount
+        if clip_amount > 0x1F {
+            reg = set_bits!(DLA_PP_CLIP_OFFSET, DLA_PP_CLIP_BITMASK, reg, 0x1F);
+        } else {
+            reg = set_bits!(DLA_PP_CLIP_OFFSET, DLA_PP_CLIP_BITMASK, reg, clip_amount);
+        }
+        self.write_reg(DLA_PP_CTRL, reg)
+    }
+
+    pub fn set_pp_rounding(&self, enable: bool) {
+        let mut reg = self.read_reg(DLA_PP_CTRL);
+        reg = set_bits!(
+            DLA_ROUNDING_OFFSET,
+            DLA_ROUNDING_BITMASK,
+            reg,
+            enable as usize
+        );
+        self.write_reg(DLA_PP_CTRL, reg);
+    }
+
+    pub fn is_ready(&self) -> bool {
+        let status = self.read_reg(DLA_STATUS_ADDR);
+        return !get_bits!(DLA_BUF_DONE_BITMASK, status) != 0;
+    }
+
+    pub fn set_bias_addr(&self, addr: u32) {
```

Comment:
Address is `u32`. Is this correct or should it be platform-specific `usize`? I.e., can it be 64-bits when dispatched by HPC?

Reply:
This is an interesting case study. We are indeed recording an _address_ in a register using a concrete, specific value. The right type here depends on the actual hardware capability, I think that should be `u64`.